pkg/ingester/transfer.go

```diff
@@ -446,32 +446,58 @@ func (i *Ingester) transferOut(ctx context.Context) error {
 }
 
 func (i *Ingester) v2TransferOut(ctx context.Context) error {
-	if len(i.TSDBState.dbs) == 0 {
-		level.Info(util.Logger).Log("msg", "nothing to transfer")
+	// Skip TSDB transfer if there are no DBs
+	i.userStatesMtx.RLock()
+	skip := len(i.TSDBState.dbs) == 0
+	i.userStatesMtx.RUnlock()
+
+	if skip {
+		level.Info(util.Logger).Log("msg", "the ingester has nothing to transfer")
 		return nil
 	}
 
-	// Close all user databases
-	wg := &sync.WaitGroup{}
-	// Only perform a shutdown once
-	once.Do(func() {
+	// This transfer function may be called multiple times in case of error,
+	// until the max number of retries is reached. For this reason, we run
+	// some initialization only once.
+	i.TSDBState.transferOnce.Do(func() {
+		// In order to transfer TSDB WAL without closing the TSDB itself - which is a
+		// pre-requisite to continue serving read requests while transferring - we need
+		// to make sure no more series will be written to the TSDB. For this reason, we
+		// wait until all in-flight write requests have been completed. No new write
+		// requests will be accepted because the "stopped" flag has already been set.
+		level.Info(util.Logger).Log("msg", "waiting for in-flight write requests to complete")
+
+		if !util.WaitGroupWithTimeout(&i.TSDBState.inflightWriteReqs, 10*time.Second) {
+			level.Warn(util.Logger).Log("msg", "timeout expired while waiting in-flight write requests to complete, transfer will continue anyway")
```

Comment:
I don't understand this decision. On one hand, you have found that transferring TSDB is only safe when there are no ongoing writes, on the other hand, this code continues even if there are still writes in progress. Wouldn't it be safer to just error out in such case?

(When using context, it may also be done/cancelled by now)

Reply:
> Wouldn't it be safer to just error out in such case?

From my perspective, in the unlikely event that the in-flight requests will take long time (or will never go down to zero) it more important to transfer all the data (and loose data points pushed by in-flight requests) than no transfer anything at all. The very last thing we want is having an ingester leaving without having transferred its state.